### PR TITLE
extract staged aggregators (with bug fixes)

### PR DESCRIPTION
**Pull request by catoverdrive**
This is basically #6534 except with some bugfixes. Stacked on #6535.

I opened a new PR because I created this branch to do some benchmarking against @chrisvittal's combiner pipeline; running `summarize` on the mt he provided (which I believe is chr22 and 10 samples wide) yields about the same runtime as the unstaged version, currently---about 30 seconds on my laptop to read/summarize/write.

@patrick-schultz @tpoterba @cseed

### CODE REVIEW

**Comment by catoverdrive** [comment]
I swapped out the read/write for a serialize/deserialize thing; got some times for a pipeline that was just read -> densify -> write on the same matrix table and got the following times:

old: [19.249044491007226, 20.367718594003236, 18.515784285002155, 18.116745114995865, 18.104985954996664]
new: [15.649361574003706, 16.205813756998396, 17.644299295003293, 16.90835837100167, 17.12396009400254]

I'd be interested in trying this on a wider matrix table, since I think that's when we start running into issues.

File: hail/src/main/scala/is/hail/expr/ir/IR.scala [hunk 1]
```diff
@@ -298,11 +298,8 @@ final case class ResultOp2(startIdx: Int, aggSigs: IndexedSeq[AggSignature]) ext
 
 final case class WriteAggs(startIdx: Int, path: IR, spec: CodecSpec, aggSigs: IndexedSeq[AggSignature]) extends IR
 final case class ReadAggs(startIdx: Int, path: IR, spec: CodecSpec, aggSigs: IndexedSeq[AggSignature]) extends IR
-
-//final case class WriteAggs(path: IR, spec: CodecSpec) extends IR
-//final case class ReadAggs(path: IR, spec: CodecSpec, aggSigs: Array[AggSignature]) extends IR
-//
-//final case class CombAggs(paths: IndexedSeq[IR], spec: CodecSpec, aggSigs: Array[AggSignature]) extends IR
+final case class SerializeAggs(startIdx: Int, serializedIdx: Int, spec: CodecSpec, aggSigs: IndexedSeq[AggSignature]) extends IR
+final case class DeserializeAggs(startIdx: Int, serializedIdx: Int, spec: CodecSpec, aggSigs: IndexedSeq[AggSignature]) extends IR
```
Comment (patrick-schultz) [inline hail/src/main/scala/is/hail/expr/ir/IR.scala]: What benefit do we get by putting (de)serialize in the IR, as opposed to just making the `FunctionWithAggRegion` trait know how to (de)serialize its aggregator state?

Comment (catoverdrive) [reply hail/src/main/scala/is/hail/expr/ir/IR.scala]: I think it's difficult to be able to have the flexibility w.r.t the codec spec that we're using to serialize---that needs to be specified at compile time. We could say that we're only ever going to use one predetermined spec to serialize state, and then precompile methods to serialize all the states and have the ability to pass in an input/output stream and specify which aggregators to serialize.

**Comment by catoverdrive** [comment]
I'm not totally happy with this yet, but it does appear to be running noticeably faster on the smaller samples I've been running on my laptop, so I've started breaking this up into actual PRs.

Some times from last week:

```
      size    |    new    |    old 
  10 x  25000 |  0m 02.72 |  0m 01.23 
  10 x 500000 |  0m 03.69 |  0m 05.38 
 200 x  25000 |  0m 01.27 |  0m 06.18 
 200 x 500000 |  0m 27.11 |  1m 06.67 
1000 x  25000 |  0m 04.91 |  0m 29.23 
1000 x 500000 |  2m 17.17 |  5m 25.63 
5000 x  25000 |  0m 22.67 |  2m 25.31 
5000 x 500000 |~12m 09s   | 27m 11.88 
```

**Comment by catoverdrive** [comment]
Closing this since all the bits have been PR'ed separately.

**Pull request closed without merging**